src/boros_impl/op/read.h

```diff
@@ -0,0 +1,18 @@
+/* This source file is part of the boros project. */
+/* SPDX-License-Identifier: ISC */
+
+#pragma once
+
+#include "op/base.h"
+
+typedef struct {
+    Operation base;
+    int fd;
```

Comment:
You could use `base.scratch` to store the `fd`. That would scrape off 8 bytes from the type's size.